lambdas/es/indexer/index.py

```diff
@@ -1,74 +1,218 @@
+            # mapper_parsing_exception that below code won't handle
+            "etag": etag,
+            "ext": ext,
+            "event": event_type,
+            "size": size,
+            "text": text,
+            "key": key,
+            "last_modified": last_modified.isoformat(),
+            "updated": datetime.utcnow().isoformat(),
+            "version_id": version_id
+        }
+
+        body = {**body, **transform_meta(meta or {})}
+
+        body["meta_text"] = " ".join([body["meta_text"], key])
+
+        self.append_document(body)
+
+        if self.size >= QUEUE_LIMIT_BYTES:
+            self.send_all()
+
+    def append_document(self, doc):
+        """append well-formed documents (used for retry or by append())"""
+        if doc["text"]:
+            # document text dominates memory footprint; OK to neglect the
+            # small fixed size for the JSON metadata
+            self.size += min(doc["size"], DOC_LIMIT_BYTES)
+        self.queue.append(doc)
+
+    def is_empty(self):
+        """is the queue empty?"""
+        return len(self.queue) == 0
+
+    def send_all(self):
+        """flush self.queue in 1-2 bulk calls"""
+        if self.is_empty():
+            return
+        elastic_host = os.environ["ES_HOST"]
+        session = boto3.session.Session()
+        credentials = session.get_credentials().get_frozen_credentials()
+        awsauth = AWSRequestsAuth(
+            # These environment variables are automatically set by Lambda
+            aws_access_key=credentials.access_key,
+            aws_secret_access_key=credentials.secret_key,
+            aws_token=credentials.token,
+            aws_host=elastic_host,
+            aws_region=session.region_name,
+            aws_service="es"
+        )
+
+        elastic = Elasticsearch(
+            hosts=[{"host": elastic_host, "port": 443}],
+            http_auth=awsauth,
+            max_backoff=get_time_remaining(self.context),
+            # Give ES time to respond when under load
+            timeout=ELASTIC_TIMEOUT,
+            use_ssl=True,
+            verify_certs=True,
+            connection_class=RequestsHttpConnection
+        )
+
+        _, errors = bulk_send(elastic, self.queue)
+        id_to_doc = {d["_id"]: d for d in self.queue}
+        send_again = []
+        for error in errors:
+            # only retry index call errors, not delete errors
+            if "index" in error:
+                inner = error["index"]
+                info = inner.get("error")
+                if isinstance(info, dict):
+                    if "mapper_parsing_exception" in info.get("type", ""):
+                        doc = id_to_doc[inner["id_"]]
+                        # zero out structured metadata and try again
+                        doc["user_meta"] = doc["system"] = {}
+                        send_again.append(doc)
+        # we won't retry after this (elasticsearch might retry 429s tho)
+        bulk_send(elastic, send_again)
```

Comment:
I think this will make a call to elastic even if `send_again` is empty? Probably good to wrap this (and the code above) in `if errors:`.